dbms/src/Interpreters/SemiJoinHelper.cpp

```diff
@@ -141,47 +143,34 @@ bool SemiJoinResult<KIND, All>::checkExprResult(
     return false;
 }
 
-template <ASTTableJoin::Kind KIND, typename Mapped>
-SemiJoinHelper<KIND, Mapped>::SemiJoinHelper(
-    Block & block_,
-    size_t left_columns_,
-    const std::vector<size_t> & right_column_indices_to_added_,
+template <ASTTableJoin::Kind KIND, ASTTableJoin::Strictness STRICTNESS, typename Maps>
+SemiJoinHelper<KIND, STRICTNESS, Maps>::SemiJoinHelper(
+    size_t input_rows_,
     size_t max_block_size_,
     const JoinNonEqualConditions & non_equal_conditions_,
     CancellationHook is_cancelled_)
-    : block(block_)
-    , left_columns(left_columns_)
-    , right_column_indices_to_add(right_column_indices_to_added_)
+    : input_rows(input_rows_)
     , max_block_size(max_block_size_)
     , is_cancelled(is_cancelled_)
     , non_equal_conditions(non_equal_conditions_)
 {
     static_assert(KIND == Semi || KIND == Anti || KIND == LeftOuterAnti || KIND == LeftOuterSemi);
-
-    right_columns = right_column_indices_to_add.size();
-    RUNTIME_CHECK(block.columns() == left_columns + right_columns);
-
-    if constexpr (KIND == LeftOuterAnti || KIND == LeftOuterSemi)
-    {
-        /// The last column is `match_helper`.
-        right_columns -= 1;
-    }
-
-    RUNTIME_CHECK(right_columns > 0);
-    RUNTIME_CHECK(non_equal_conditions.other_cond_expr != nullptr);
 }
 
-template <ASTTableJoin::Kind KIND, typename Mapped>
-void SemiJoinHelper<KIND, Mapped>::joinResult(std::list<Result *> & res_list)
+template <ASTTableJoin::Kind KIND, ASTTableJoin::Strictness STRICTNESS, typename Maps>
+void SemiJoinHelper<KIND, STRICTNESS, Maps>::doJoin()
 {
+    if constexpr (STRICTNESS == ASTTableJoin::Strictness::Any)
+    {
+        return;
+    }
+    assert(!probe_res_list.empty());
```

Comment:
Add a `if constexpr (STRICTNESS == ASTTableJoin::Strictness::All)` to cover the whole function and then some useless functions like `getRowNum` in `SemiJoinResult<KIND, Any>` can be removed.